File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/I._Разные_деревья_поиска/solution.py
import math
import sys

def solution() -> None:
    """
    Time Complexity O(1)
    Memory Complexity O(1)
    """
    n: int = int(sys.stdin.readline().rstrip())
    catalan_number: int = math.factorial(2 * n) // (math.factorial(n + 1) * math.factorial(n))
    sys.stdout.write(f'{catalan_number}\n')


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/I._Разные_деревья_поиска/solution.cpp
#include <iostream>
#include <vector>

long long countBST(int n) {
    std::vector<long long> catalan(n + 1, 0);
    catalan[0] = 1;
    for (int i = 1; i <= n; ++i) {
        for (int j = 0; j < i; ++j) {
            catalan[i] += catalan[j] * catalan[i - 1 - j];
        }
    }
    return catalan[n];
}

int main() {
    /*
    Time Complexity O(1)
    Memory Complexity O(1)
    */
    int n;
    std::cin >> n;
    std::cout << countBST(n) << std::endl;
}